### PR TITLE
Refactor and improve fusion

### PR DESCRIPTION
**Pull request by hhorii**
### Summary
This PR refactors and improves fusion optimization.

### Details and comments
`Transpile::Fusion` generates a matrix. `matrix_sequence` is deprecated from `OpType`. 
`QV::QubitVector` specializes some 2qubit operations with matrixes.

### CODE REVIEW

File: src/simulators/statevector/qubitvector.hpp [hunk 1]
```diff
@@ -1801,150 +1709,102 @@ void QubitVector<data_t>::apply_diagonal_matrix(const uint_t qubit,
...
-      u[(i + low_delta + high_delta)  + (i + high_delta)             * dst_dim] = sorted_vmat[3 + 2 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[3 + 3 * src_dim];
-
-      filled[i] = true;
-      filled[i + low_delta] = true;
-      filled[i + high_delta] = true;
-      filled[i + low_delta + high_delta] = true;
-    }
-    //TODO: } else if (src_qubits.size() == 3) {
+  if (mat[0] == 1.0 && mat[4] == 0.0 && mat[8]  == 0.0 && mat[12] == 0.0
```
Comment (atilag) [inline src/simulators/statevector/qubitvector.hpp]: Ok, this is very tought to read.
Can we make a helper function/lambda that returns `true` or `false` to add some meaning to this conditional?
Something like, I don't know:
```c++
if( isWhateverMeangfullMatrixFormName(mat) ){ 
```

File: src/simulators/statevector/qubitvector.hpp [hunk 2]
```diff
@@ -1801,150 +1709,102 @@ void QubitVector<data_t>::apply_diagonal_matrix(const uint_t qubit,
...
+    auto lambda = [&](const areg_t<4> &inds, const cvector_t &_mat)->void {
+      std::array<complex_t, 4> cache;
+      for (size_t i = 1; i < 4; i++)
+        cache[i] = data_[inds[i]];
+      data_[inds[1]] = _mat[13] * cache[3];
+      data_[inds[2]] = _mat[10] * cache[2];
+      data_[inds[3]] = cache[1];
+    };
+    apply_lambda(lambda, areg_t<2>( { { qubit0, qubit1 } }), mat);
+  } else if (mat[0] == 1.0 && mat[4] == 0.0 && mat[8]  == 0.0 && mat[12] == 0.0
```
Comment (atilag) [inline src/simulators/statevector/qubitvector.hpp]: Same as above here

File: src/simulators/statevector/qubitvector.hpp [hunk 3]
```diff
@@ -1801,150 +1709,102 @@ void QubitVector<data_t>::apply_diagonal_matrix(const uint_t qubit,
...
+    auto lambda = [&](const areg_t<4> &inds, const cvector_t &_mat)->void {
+      std::array<complex_t, 4> cache;
+      for (size_t i = 1; i < 4; i++)
+        cache[i] = data_[inds[i]];
+      data_[inds[1]] = _mat[5] * cache[1];
+      data_[inds[2]] = _mat[14] * cache[3];
+      data_[inds[3]] = cache[2];
+    };
+    apply_lambda(lambda, areg_t<2>( { { qubit0, qubit1 } }), mat);
+  } else if (mat[0] == 1.0 && mat[4] == 0.0 && mat[8]  == 0.0 && mat[12] == 0.0
```
Comment (atilag) [inline src/simulators/statevector/qubitvector.hpp]: Same here

File: src/simulators/statevector/qubitvector.hpp [hunk 4]
```diff
@@ -1801,150 +1709,102 @@ void QubitVector<data_t>::apply_diagonal_matrix(const uint_t qubit,
...
+    auto lambda = [&](const areg_t<4> &inds, const cvector_t &_mat)->void {
+      std::array<complex_t, 4> cache;
+      for (size_t i = 1; i < 4; i++)
+      cache[i] = data_[inds[i]];
+      data_[inds[1]] = _mat[13] * cache[3];
+      data_[inds[2]] = cache[2];
+      data_[inds[3]] = _mat[7] * cache[1];
+    };
+    apply_lambda(lambda, areg_t<2>( { { qubit0, qubit1 } }), mat);
+  } else if (mat[0] == 1.0 && mat[4] == 0.0 && mat[8]  == 0.0 && mat[12] == 0.0
```
Comment (atilag) [inline src/simulators/statevector/qubitvector.hpp]: Same here

File: src/simulators/statevector/qubitvector.hpp [hunk 5]
```diff
@@ -1801,150 +1709,102 @@ void QubitVector<data_t>::apply_diagonal_matrix(const uint_t qubit,
...
+    auto lambda = [&](const areg_t<4> &inds, const cvector_t &_mat)->void {
+      std::array<complex_t, 4> cache;
+      for (size_t i = 1; i < 4; i++)
+      cache[i] = data_[inds[i]];
+      data_[inds[1]] = cache[1];
+      data_[inds[2]] = _mat[14] * cache[3];
+      data_[inds[3]] = _mat[11] * cache[2];
+    };
+    apply_lambda(lambda, areg_t<2>( { { qubit0, qubit1 } }), mat);
+  } else if (mat[0] == 1.0 && mat[4] == 0.0 && mat[8]  == 0.0 && mat[12] == 0.0
```
Comment (atilag) [inline src/simulators/statevector/qubitvector.hpp]: Same here

File: src/transpile/fusion.hpp [hunk 6]
```diff
@@ -29,7 +29,7 @@ using reg_t = std::vector<uint_t>;
 
 class Fusion : public CircuitOptimization {
```
Comment (atilag) [inline src/transpile/fusion.hpp]: We need some code documentation here.
What is `Fusion`?

Comment (atilag) [reply src/transpile/fusion.hpp]: In general, we have to document every method in this class

File: src/transpile/fusion.hpp [hunk 7]
```diff
@@ -149,56 +159,40 @@ void Fusion::optimize_circuit(Circuit& circ,
...
-        ret = true;
-        oplist_t optimized_buffer = aggregate(buffer);
-        optimized_ops.insert(optimized_ops.end(), optimized_buffer.begin(), optimized_buffer.end());
-        buffer.clear();
-      }
-      optimized_ops.push_back(op);
-    } else {
-      buffer.push_back(op);
+    if (!can_apply_fusion(circ.ops[op_idx])) {
+      if (fusion_start != op_idx)
```
Comment (atilag) [inline src/transpile/fusion.hpp]: When nesting non-trivial conditionals, it's better if we add curly braces to all the nestings, makes code easier to follow

File: src/transpile/fusion.hpp [hunk 8]
```diff
@@ -228,83 +222,208 @@ bool Fusion::can_apply_fusion(const op_t& op) const {
...
+      filled[i] = true;
+      filled[i + low_delta] = true;
+      filled[i + high_delta] = true;
+      filled[i + low_delta + high_delta] = true;
+    }
+    //TODO: } else if (src_qubits.size() == 3) {
+  } else {
+    std::stringstream ss;
+    ss << "Fusion::illegal qubit number:" << src_qubits.size();
+    throw std::runtime_error(ss.str());
```
Comment (atilag) [inline src/transpile/fusion.hpp]: We want to fail as fast as possible, so I'd move this `else` to the begginning of the function and write something like:
```c++
cmatrix_t Fusion::expand_matrix(const reg_t& src_qubits, const reg_t& dst_sorted_qubits, const cmatrix_t& mat) const {

if( src_qubits.size() > 2 )
    throw std::runtime_error("Fusion::illegal qubit number: " + std::to_string(src_qubits.size()));
```
There's no need for a `stringstream`.

File: src/transpile/fusion.hpp [hunk 9]
```diff
@@ -432,7 +563,9 @@ cmatrix_t Fusion::matrix(const op_t& op) const {
   } else if (op.type == optype_t::matrix) {
     return op.mats[0];
   } else {
-    throw std::runtime_error("Fusion: unexpected operation type");
+    std::stringstream msg;
+    msg << "Fusion: unexpected operation type:" << op.type;
+    throw std::runtime_error(msg.str());
```
Comment (atilag) [inline src/transpile/fusion.hpp]:
```c++
 throw std::runtime_error("Fusion: unexpected operation type:" + std::to_string(op.type));
```

File: src/framework/operations.hpp [hunk 10]
```diff
@@ -36,7 +36,7 @@ enum class RegComparison {Equal, NotEqual, Less, LessEqual, Greater, GreaterEqua
 // Enum class for operation types
 enum class OpType {
   gate, measure, reset, bfunc, barrier, snapshot,
-  matrix, matrix_sequence, multiplexer, kraus, roerror, noise_switch, initialize
+  matrix, multiplexer, kraus, roerror, noise_switch, initialize, nop
```
[the remark this answers is not in the document]
Comment (hhorii) [reply src/framework/operations.hpp]: in transpile, I do not want to erase elements in `ops`. `nop` means no operation for simulator.

File: src/simulators/statevector/qubitvector.hpp (hunk 1 above)
Comment (hhorii) [reply src/simulators/statevector/qubitvector.hpp]: Hummm, a bit difficult to express these kinds of matrixes...

File: src/framework/operations.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (hhorii) [reply src/framework/operations.hpp]: I removed `nop` in 750baf4.

File: src/framework/operations.hpp [hunk 11]
```diff
@@ -390,14 +425,25 @@ inline Op make_unitary(const reg_t &qubits, const cmatrix_t &mat, std::string la
...
+  op.type = OpType::matrix;
+  op.name = "fusion";
+  op.qubits = qubits;
+  op.mats = {mat};
   if (label != "")
     op.string_params = {label};
+
+#ifdef DEBUG
+  for (const Op& fusioned_op: fusioned_ops) {
+    op.fusioned_types.push_back(fusioned_op.type);
```
[the remark this answers is not in the document]
Comment (hhorii) [reply src/framework/operations.hpp]: I removed it.

File: src/framework/utils.hpp [hunk 12]
```diff
@@ -485,6 +487,15 @@ cvector_t VMatrix::u3(double theta, double phi, double lambda) {
 // Implementations: Matrix functions
 //==============================================================================
 
+template<class T>
+void initialize_matrix(const matrix<T>& mat, T v) {
+  std::vector<T> vec;
```
[the remark this answers is not in the document]
Comment (hhorii) [reply src/framework/utils.hpp]: I removed this.

File: src/simulators/statevector/qubitvector.hpp (hunk 1 above)
Comment (hhorii) [reply src/simulators/statevector/qubitvector.hpp]: I removed specialization for 2qubit matrix operation. diagonal matrix operation covers all.

File: src/simulators/statevector/qubitvector.hpp (hunk 2 above)
Comment (hhorii) [reply src/simulators/statevector/qubitvector.hpp]: I removed specialization for 2qubit matrix operation. diagonal matrix operation covers all.

File: src/simulators/statevector/qubitvector.hpp (hunk 4 above)
Comment (hhorii) [reply src/simulators/statevector/qubitvector.hpp]: Same as above.

File: src/simulators/statevector/qubitvector.hpp (hunk 5 above)
Comment (hhorii) [reply src/simulators/statevector/qubitvector.hpp]: same the above.

File: src/transpile/fusion.hpp (hunk 9 above)
Comment (hhorii) [reply src/transpile/fusion.hpp]: I have changed codes as you pointed. Thanks.

File: src/framework/utils.hpp [hunk 13]
```diff
@@ -510,6 +510,15 @@ std::vector<T> vectorize_matrix(const matrix<T>& mat) {
   return vec;
 }
 
+template<class T>
+std::vector<T> vectorize_diagonal_matrix(const matrix<T>& mat) {
```
Comment (chriseclectic) [inline src/framework/utils.hpp]: This function actually just extracts the diagonal from a square matrix, so maybe we should call it `matrix_diagonal`

**Pull request merged**